File: cpp/1038.binary-search-tree-to-greater-sum-tree.cpp
/*
 * @lc app=leetcode id=1038 lang=cpp
 *
 * [1038] Binary Search Tree to Greater Sum Tree
 */
#include <vector>
using namespace std;
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// t:O(N), s:O(N)
class Solution
{
public:
    TreeNode *bstToGst(TreeNode *root)
    {
        vector<TreeNode *> st;
        auto cur = root;
        int accum_val = 0;
        // inorder traversal
        while (true)
        {
            if (cur)
            {
                st.push_back(cur);
                cur = cur->right;
            }
            else if (!st.empty())
            {
                accum_val = st.back()->val += accum_val;
                cur = st.back()->left;
                st.pop_back();
            }
            else
            {
                break;
            }
        }
        return root;
    }
};
// @lc code=end

File: py/1038.binary-search-tree-to-greater-sum-tree.py
#
# @lc app=leetcode id=1038 lang=python3
#
# [1038] Binary Search Tree to Greater Sum Tree
#
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# t:O(N), s:O(N)
class Solution:
    def bstToGst(self, root: TreeNode) -> TreeNode:
        st: List[TreeNode] = []
        cur = root
        accum_val = 0
        while cur or len(st):
            if cur:
                st.append(cur)
                cur = cur.right
            else:
                st[-1].val += accum_val
                accum_val = st[-1].val
                cur = st.pop().left
        return root
# @lc code=end
